### ISSUE TITLE
Custom Expression using `case()` function fails when referencing the same column names

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#14854]
**Describe the bug**
Custom Expression using `case()` function fails when referencing the same column names on different tables in true-and-else statement.

**To Reproduce**
1. Custom question > Sample Dataset > Orders
2. Custom Column - `case([Discount] > 0, [Created At], [Product → Created At])` as "CE"
![image](https://user-images.githubusercontent.com/1447303/108107767-f6e88580-708f-11eb-9008-278b6ff7e69f.png)
3. Query fails with `Ambiguous column name "CREATED_AT";`
![image](https://user-images.githubusercontent.com/1447303/108108353-d10fb080-7090-11eb-8268-e28ea98824a3.png)
<details><summary>Full stacktrace</summary>

```
2021-02-16 19:46:49,324 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2021-02-16T19:46:46.598901+01:00[Europe/Copenhagen]",
 :state "90059",
 :json_query
 {:database 4,
  :query {:source-table 11, :expressions {:CE ["case" [[[">" ["field-id" 81] 0] ["field-id" 86]]] {:default ["fk->" ["field-id" 82] ["field-id" 108]]}]}},
  :type "query",
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native
 {:query
  "SELECT \"source\".\"ID\" AS \"ID\", \"source\".\"USER_ID\" AS \"USER_ID\", \"source\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"source\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"source\".\"TAX\" AS \"TAX\", \"source\".\"TOTAL\" AS \"TOTAL\", \"source\".\"DISCOUNT\" AS \"DISCOUNT\", \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"QUANTITY\" AS \"QUANTITY\", \"source\".\"CE\" AS \"CE\" FROM (SELECT \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" AS \"DISCOUNT\", \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" AS \"CREATED_AT\", \"PRODUCTS__via__PRODUCT_ID\".\"CREATED_AT\" AS \"PRODUCTS__via__PRODUCT_ID__CREATED_AT\", \"PUBLIC\".\"ORDERS\".\"ID\" AS \"ID\", \"PUBLIC\".\"ORDERS\".\"USER_ID\" AS \"USER_ID\", \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"PUBLIC\".\"ORDERS\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"PUBLIC\".\"ORDERS\".\"TAX\" AS \"TAX\", \"PUBLIC\".\"ORDERS\".\"TOTAL\" AS \"TOTAL\", \"PUBLIC\".\"ORDERS\".\"QUANTITY\" AS \"QUANTITY\", CASE WHEN \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" > 0 THEN \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" ELSE \"CREATED_AT\" END AS \"CE\", \"PRODUCTS__via__PRODUCT_ID\".\"ID\" AS \"PRODUCTS__via__PRODUCT_ID__ID\" FROM \"PUBLIC\".\"ORDERS\" LEFT JOIN \"PUBLIC\".\"PRODUCTS\" \"PRODUCTS__via__PRODUCT_ID\" ON \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" = \"PRODUCTS__via__PRODUCT_ID\".\"ID\") \"source\" LIMIT 2000",
  :params nil},
 :status :failed,
 :class org.h2.jdbc.JdbcSQLException,
 :stacktrace
 ["org.h2.message.DbException.getJdbcSQLException(DbException.java:357)"
  "org.h2.message.DbException.get(DbException.java:179)"
  "org.h2.message.DbException.get(DbException.java:155)"
  "org.h2.expression.ExpressionColumn.mapColumn(ExpressionColumn.java:127)"
  "org.h2.expression.ExpressionColumn.mapColumns(ExpressionColumn.java:97)"
  "org.h2.expression.Function.mapColumns(Function.java:1992)"
  "org.h2.expression.Alias.mapColumns(Alias.java:46)"
  "org.h2.command.dml.Select.mapColumns(Select.java:1237)"
  "org.h2.command.dml.Select.init(Select.java:826)"
  "org.h2.command.Parser.readTableFilter(Parser.java:1359)"
  "org.h2.command.Parser.parseSelectSimpleFromPart(Parser.java:2138)"
  "org.h2.command.Parser.parseSelectSimple(Parser.java:2287)"
  "org.h2.command.Parser.parseSelectSub(Parser.java:2133)"
  "org.h2.command.Parser.parseSelectUnion(Parser.java:1946)"
  "org.h2.command.Parser.parseSelect(Parser.java:1919)"
  "org.h2.command.Parser.parsePrepared(Parser.java:463)"
  "org.h2.command.Parser.parse(Parser.java:335)"
  "org.h2.command.Parser.parse(Parser.java:307)"
  "org.h2.command.Parser.prepareCommand(Parser.java:278)"
  "org.h2.engine.Session.prepareLocal(Session.java:611)"
  "org.h2.engine.Session.prepareCommand(Session.java:549)"
  "org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1247)"
  "org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:76)"
  "org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:1135)"
  "com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:267)"
  "--> driver.sql_jdbc.execute$fn__77429.invokeStatic(execute.clj:236)"
  "driver.sql_jdbc.execute$fn__77429.invoke(execute.clj:234)"
  "driver.sql_jdbc.execute$prepared_statement_STAR_.invokeStatic(execute.clj:254)"
  "driver.sql_jdbc.execute$prepared_statement_STAR_.invoke(execute.clj:251)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:387)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:374)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:383)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:374)"
  "driver.sql_jdbc$fn__79015.invokeStatic(sql_jdbc.clj:54)"
  "driver.sql_jdbc$fn__79015.invoke(sql_jdbc.clj:52)"
  "driver.h2$fn__77819.invokeStatic(h2.clj:81)"
  "driver.h2$fn__77819.invoke(h2.clj:78)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:68)"
  "query_processor.context.default$default_runf.invoke(default.clj:66)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
  "query_processor.reducible$pivot.invoke(reducible.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__46355.invoke(mbql_to_native.clj:25)"
  "query_processor.middleware.check_features$check_features$fn__45631.invoke(check_features.clj:41)"
  "query_processor.middleware.limit$limit$fn__46341.invoke(limit.clj:37)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__45282.invoke(cache.clj:211)"
  "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__46520.invoke(optimize_datetime_filters.clj:133)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__44428.invoke(auto_parse_filter_values.clj:43)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__40809.invoke(wrap_value_literals.clj:147)"
  "query_processor.middleware.annotate$add_column_info$fn__40672.invoke(annotate.clj:582)"
  "query_processor.middleware.permissions$check_query_permissions$fn__45506.invoke(permissions.clj:69)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__47043.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45704.invoke(cumulative_aggregations.clj:60)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__47356.invoke(resolve_joined_fields.clj:94)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__47661.invoke(resolve_joins.clj:178)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__44018.invoke(add_implicit_joins.clj:181)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__46316.invoke(large_int_id.clj:44)"
  "query_processor.middleware.format_rows$format_rows$fn__46296.invoke(format_rows.clj:74)"
  "query_processor.middleware.desugar$desugar$fn__45770.invoke(desugar.clj:21)"
  "query_processor.middleware.binning$update_binning_strategy$fn__44788.invoke(binning.clj:228)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__45306.invoke(resolve_fields.clj:24)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__43648.invoke(add_dimension_projections.clj:316)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__43879.invoke(add_implicit_clauses.clj:146)"
  "query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__48091.invoke(upgrade_field_literals.clj:45)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__44175.invoke(add_source_metadata.clj:124)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__47240.invoke(reconcile_breakout_and_order_by_bucketing.clj:97)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__44375.invoke(auto_bucket_datetimes.clj:139)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__45353.invoke(resolve_source_table.clj:45)"
  "query_processor.middleware.parameters$substitute_parameters$fn__47025.invoke(parameters.clj:111)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__45405.invoke(resolve_referenced.clj:79)"
  "query_processor.middleware.expand_macros$expand_macros$fn__46026.invoke(expand_macros.clj:155)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__44184.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__48027.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__47251$fn__47255.invoke(resolve_database_and_driver.clj:31)"
  "driver$do_with_driver.invokeStatic(driver.clj:60)"
  "driver$do_with_driver.invoke(driver.clj:56)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__47251.invoke(resolve_database_and_driver.clj:25)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__46244.invoke(fetch_source_query.clj:264)"
  "query_processor.middleware.store$initialize_store$fn__48036$fn__48037.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__48036.invoke(store.clj:10)"
  "query_processor.middleware.validate$validate_query$fn__48098.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__46368.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__44036.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__48012.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__45647.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__47114.invoke(process_userland_query.clj:135)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__45590.invoke(catch_exceptions.clj:173)"
  "query_processor.reducible$async_qp$qp_STAR___37450$thunk__37451.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___37450.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___37459$fn__37462.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___37459.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:237)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:233)"
  "query_processor$fn__48144$process_query_and_save_execution_BANG___48153$fn__48156.invoke(query_processor.clj:249)"
  "query_processor$fn__48144$process_query_and_save_execution_BANG___48153.invoke(query_processor.clj:241)"
  "query_processor$fn__48188$process_query_and_save_with_max_results_constraints_BANG___48197$fn__48200.invoke(query_processor.clj:261)"
  "query_processor$fn__48188$process_query_and_save_with_max_results_constraints_BANG___48197.invoke(query_processor.clj:254)"
  "api.dataset$fn__54298$fn__54301.invoke(dataset.clj:55)"
  "query_processor.streaming$streaming_response_STAR_$fn__54279$fn__54280.invoke(streaming.clj:72)"
  "query_processor.streaming$streaming_response_STAR_$fn__54279.invoke(streaming.clj:71)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
  "async.streaming_response$do_f_async$fn__16055.invoke(streaming_response.clj:84)"],
 :context :ad-hoc,
 :error
 "Ambiguous column name \"CREATED_AT\"; SQL statement:\n-- Metabase:: userID: 1 queryType: MBQL queryHash: 6d78f8687f546497906b30c598cd72454ea3a511461d6bd99bccd01998ddfd6c\nSELECT \"source\".\"ID\" AS \"ID\", \"source\".\"USER_ID\" AS \"USER_ID\", \"source\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"source\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"source\".\"TAX\" AS \"TAX\", \"source\".\"TOTAL\" AS \"TOTAL\", \"source\".\"DISCOUNT\" AS \"DISCOUNT\", \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"QUANTITY\" AS \"QUANTITY\", \"source\".\"CE\" AS \"CE\" FROM (SELECT \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" AS \"DISCOUNT\", \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" AS \"CREATED_AT\", \"PRODUCTS__via__PRODUCT_ID\".\"CREATED_AT\" AS \"PRODUCTS__via__PRODUCT_ID__CREATED_AT\", \"PUBLIC\".\"ORDERS\".\"ID\" AS \"ID\", \"PUBLIC\".\"ORDERS\".\"USER_ID\" AS \"USER_ID\", \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"PUBLIC\".\"ORDERS\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"PUBLIC\".\"ORDERS\".\"TAX\" AS \"TAX\", \"PUBLIC\".\"ORDERS\".\"TOTAL\" AS \"TOTAL\", \"PUBLIC\".\"ORDERS\".\"QUANTITY\" AS \"QUANTITY\", CASE WHEN \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" > 0 THEN \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" ELSE \"CREATED_AT\" END AS \"CE\", \"PRODUCTS__via__PRODUCT_ID\".\"ID\" AS \"PRODUCTS__via__PRODUCT_ID__ID\" FROM \"PUBLIC\".\"ORDERS\" LEFT JOIN \"PUBLIC\".\"PRODUCTS\" \"PRODUCTS__via__PRODUCT_ID\" ON \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" = \"PRODUCTS__via__PRODUCT_ID\".\"ID\") \"source\" LIMIT 2000 [90059-197]",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 4,
  :query
  {:source-table 11,
   :expressions
   {:CE
    [:case
     [[[:> [:field-id 81] [:value 0 {:base_type :type/Float, :special_type :type/Discount, :database_type "DOUBLE", :name "DISCOUNT"}]] [:field-id 86]]]
     {:default [:joined-field "PRODUCTS__via__PRODUCT_ID" [:field-id 108]]}]},
   :fields [[:field-id 83] [:field-id 80] [:field-id 82] [:field-id 84] [:field-id 87] [:field-id 88] [:field-id 81] [:datetime-field [:field-id 86] :default] [:field-id 85] [:expression "CE"]],
   :joins [{:strategy :left-join, :alias "PRODUCTS__via__PRODUCT_ID", :condition [:= [:field-id 82] [:joined-field "PRODUCTS__via__PRODUCT_ID" [:field-id 105]]], :source-table 10, :fk-field-id 82}],
   :limit 2000},
  :type :query,
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash [109, 120, -8, 104, 127, 84, 100, -105, -112, 107, 48, -59, -104, -51, 114, 69, 78, -93, -91, 17, 70, 29, 107, -39, -101, -52, -48, 25, -104, -35, -3, 108]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}

2021-02-16 19:46:49,357 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 2.8 s (23 DB calls) App DB connections: 0/4 Jetty threads: 3/50 (6 idle, 0 queued) (51 total active threads) Queries in flight: 0 (0 queued)
```

</details>

**Information about your Metabase Installation:**
The used to work on 0.35.0 thru 0.38.0-rc5, but broken on latest `release-x.38.x` (tested commit `dc50e8a`)

**Additional context**
Related to #13122 (test repro was being created for that, when this was discovered)

**Comment by camsaul** [comment]
This might be a case of the FE using `field-literal` to refer to the column instead of whatever comes back from `field_ref` (e.g. `field-id`)... I'd have to see the generated MBQL tho to determine whether that's the case or not

**Comment by nemanjaglumac** [comment]
@camsaul does this help?

```javascript
dataset_query: {
  type: "query",
  query: {
    "source-table": 2,
    expressions: {
      CE: [
        "case",
        [[[">", ["field-id", 9], 0], ["field-id", 12]]],
        { default: ["fk->", ["field-id", 13], ["field-id", 7]] },
      ],
    },
  },
  database: 1,
},
```

**Comment by camsaul** [comment]
@nemanjaglumac yes, thanks.

That does look correctly actually... it's not using `field-literal`. So this would probably be a backend fix then.

**Comment by flamber** [comment]
All my issues have full stacktraces - usually collapsed in a `<details>`, since it otherwise makes viewing the issue difficult.

**Comment by camsaul** [comment]
@flamber awesome. I didn't notice that. Thanks

**Comment by camsaul** [comment]
Fixed by #14858
